File: src/datamart.py
from pyspark.sql import DataFrame, SparkSession, SQLContext


class DataMart:
    def __init__(self, spark: SparkSession):
        self.spark = spark
        self.spark_context = spark.sparkContext
        self.sql_context = SQLContext(self.spark_context, spark)
        self.jwm_datamart = self.spark_context._jvm.DataMart(spark._jsparkSession)

    def read_dataset(self, path: str, sep: str = ",") -> DataFrame:
        jvm_data = self.jwm_datamart.readAndProcessDataset(path, sep)
        return DataFrame(jvm_data, self.sql_context)

    def write_predictions(self, df: DataFrame):
        self.jwm_datamart.writePredictions(df._jdf)

    def read_database(self, keyspace: str, table: str) -> DataFrame:
        jvm_data = self.jwm_datamart.read(keyspace, table)
        return DataFrame(jvm_data, self.sql_context)
    
    def write_data(self, data: DataFrame):
        self.jwm_datamart.writeData(data._jdf)

File: src/datamart/src/main/scala/DataMart.scala
import org.apache.spark.sql.{SparkSession, DataFrame}
import org.apache.spark.sql.functions._
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}
import db.{DataBase}
import org.apache.spark.sql.types.{DoubleType, StructType}
import preprocess.Preprocessor

class DataMart(spark: SparkSession) {

  private val db = new DataBase(spark)

  def readAndProcessDataset(path: String, sep: String): DataFrame = {
    val df = spark.read
      .format("csv")
      .option("header", "true")
      .option("sep", sep)
      .option("inferSchema", "true")
      .load(path)

    val transforms: Seq[DataFrame => DataFrame] = Seq(
      Preprocessor.fillNa,
      Preprocessor.assembleVector,
      Preprocessor.scaleAssembledDataset
    )

    val transformed = transforms.foldLeft(df) { (df, f) => f(df) }
    transformed
  }

  def write(df: DataFrame, keyspace: String, table: String, mode: String): Unit = {
    db.writeDatababse(df, keyspace, table, mode)
  }

  def read(keyspace: String, table: String): DataFrame = {
    db.readDatabase(keyspace, table)
  }

  def writeData(df: DataFrame): Unit = {
    db.writeDatababse(df, "food", "data", "overwrite")
  }

  def writePredictions(df: DataFrame): Unit = {
    db.writeDatababse(df, "food", "predictions", "append")
  }
}
